common/cuda_hip/base/batch_multi_vector_kernels.hpp.inc

```diff
@@ -0,0 +1,272 @@
+    const int max_li = x.num_rows * x.num_rhs;
+    for (int li = threadIdx.x; li < max_li; li += blockDim.x) {
+        const int row = li / x.num_rhs;
+        const int col = li % x.num_rhs;
+
+        x.values[row * x.stride + col] =
+            alpha.values[map(col)] * x.values[row * x.stride + col];
+    }
+}
+
+template <typename ValueType, typename Mapping>
+__global__
+__launch_bounds__(default_block_size, sm_multiplier) void scale_kernel(
+    const gko::batch_multi_vector::uniform_batch<const ValueType> alpha,
+    const gko::batch_multi_vector::uniform_batch<ValueType> x, Mapping map)
+{
+    for (size_type batch_id = blockIdx.x; batch_id < x.num_batch_entries;
+         batch_id += gridDim.x) {
+        const auto alpha_b = gko::batch::batch_entry(alpha, batch_id);
+        const auto x_b = gko::batch::batch_entry(x, batch_id);
+        scale(alpha_b, x_b, map);
+    }
+}
+
+
+template <typename ValueType, typename Mapping>
+__device__ __forceinline__ void add_scaled(
+    const gko::batch_multi_vector::batch_entry<const ValueType>& alpha,
+    const gko::batch_multi_vector::batch_entry<const ValueType>& x,
+    const gko::batch_multi_vector::batch_entry<ValueType>& y, Mapping map)
+{
+    const int max_li = x.num_rows * x.num_rhs;
+    for (int li = threadIdx.x; li < max_li; li += blockDim.x) {
+        const int row = li / x.num_rhs;
+        const int col = li % x.num_rhs;
+
+        y.values[row * y.stride + col] +=
+            alpha.values[map(col)] * x.values[row * x.stride + col];
+    }
+}
+
+template <typename ValueType, typename Mapping>
+__global__
+__launch_bounds__(default_block_size, sm_multiplier) void add_scaled_kernel(
+    const gko::batch_multi_vector::uniform_batch<const ValueType> alpha,
+    const gko::batch_multi_vector::uniform_batch<const ValueType> x,
+    const gko::batch_multi_vector::uniform_batch<ValueType> y, Mapping map)
+{
+    for (size_type batch_id = blockIdx.x; batch_id < x.num_batch_entries;
+         batch_id += gridDim.x) {
+        const auto alpha_b = gko::batch::batch_entry(alpha, batch_id);
+        const auto x_b = gko::batch::batch_entry(x, batch_id);
+        const auto y_b = gko::batch::batch_entry(y, batch_id);
+        add_scaled(alpha_b, x_b, y_b, map);
+    }
+}
+
+
+template <typename Group, typename ValueType, typename Mapping>
+__device__ __forceinline__ void gen_one_dot(
+    const gko::batch_multi_vector::batch_entry<const ValueType>& x,
+    const gko::batch_multi_vector::batch_entry<const ValueType>& y,
+    const int rhs_index,
+    const gko::batch_multi_vector::batch_entry<ValueType>& result,
+    Group subgroup, Mapping conj_map)
+{
+    ValueType val = zero<ValueType>();
+
+    for (int r = subgroup.thread_rank(); r < x.num_rows; r += subgroup.size()) {
+        val += conj_map(x.values[r * x.stride + rhs_index]) *
+               y.values[r * y.stride + rhs_index];
+    }
+
+    // subgroup level reduction
+    val = reduce(subgroup, val, thrust::plus<ValueType>{});
```

Comment:
I think it makes sense to unify this ? But that is another discussion. But I would like to keep it like this because technically this is cuda/hip, so subgroup makes more sense than subwarp/subwavefront ?